File: app/src/main/java/com/sharpdroid/registroelettronico/Fragments/FragmentAgenda.java
package com.sharpdroid.registroelettronico.Fragments;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.design.widget.CoordinatorLayout;
import android.support.design.widget.Snackbar;
import android.support.v4.app.Fragment;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;

import com.github.clans.fab.FloatingActionButton;
import com.github.clans.fab.FloatingActionMenu;
import com.github.sundeepk.compactcalendarview.CompactCalendarView;
import com.sharpdroid.registroelettronico.API.SpiaggiariApiClient;
import com.sharpdroid.registroelettronico.Activities.AddEventActivity;
import com.sharpdroid.registroelettronico.Adapters.AgendaAdapter;
import com.sharpdroid.registroelettronico.BottomSheet.AgendaBS;
import com.sharpdroid.registroelettronico.Databases.RegistroDB;
import com.sharpdroid.registroelettronico.Interfaces.Client.AdvancedEvent;
import com.sharpdroid.registroelettronico.R;
import com.transitionseverywhere.ChangeText;
import com.transitionseverywhere.TransitionManager;

import org.apache.commons.lang3.text.WordUtils;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Date;
import java.util.List;
import java.util.Locale;
import java.util.TimeZone;

import butterknife.BindView;
import butterknife.ButterKnife;
import io.reactivex.android.schedulers.AndroidSchedulers;

import static com.sharpdroid.registroelettronico.Utils.Metodi.addEventToCalendar;
import static com.sharpdroid.registroelettronico.Utils.Metodi.convertEvents;
import static com.sharpdroid.registroelettronico.Utils.Metodi.eventToString;
import static com.sharpdroid.registroelettronico.Utils.Metodi.getSubjectNameOrProfessorName;

// DONE: 19/01/2017 Aggiungere eventi all'agenda
// DONE: 19/01/2017 Aggiungere eventi dell'agenda nel calendario del telefono

public class FragmentAgenda extends Fragment implements CompactCalendarView.CompactCalendarViewListener, AgendaAdapter.AgendaClickListener, AgendaBS.Listener {
    final private String TAG = FragmentAgenda.class.getSimpleName();
    SimpleDateFormat month = new SimpleDateFormat("MMMM", Locale.getDefault());
    SimpleDateFormat year = new SimpleDateFormat("yyyy", Locale.getDefault());
    @BindView(R.id.recycler)
    RecyclerView recycler;
    @BindView(R.id.place_holder)
    View place_holder;
    @BindView(R.id.coordinator_layout)
    CoordinatorLayout mCoordinatorLayout;
    @BindView(R.id.fab_big_add)
    FloatingActionMenu addFAB;
    @BindView(R.id.fab_mini_verifica)
    FloatingActionButton verificaFAB;
    @BindView(R.id.fab_mini_esercizi)
    FloatingActionButton eserciziFAB;
    @BindView(R.id.fab_mini_altro)
    FloatingActionButton altroFAB;

    private CompactCalendarView mCompactCalendarView;
    private Toolbar mToolbar;
    private Context mContext;
    private RegistroDB mRegistroDB;
    private AgendaAdapter adapter;
    private Date mDate;
    private List<AdvancedEvent> events = new ArrayList<>();


    public FragmentAgenda() {
    }

    @Override
    public View onCreateView(final LayoutInflater inflater,
                             ViewGroup container, Bundle savedInstanceState) {
        setHasOptionsMenu(true);
        mContext = getContext();
        return inflater.inflate(R.layout.fragment_agenda, container, false);
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        ButterKnife.bind(this, view);

        mToolbar = (Toolbar) getActivity().findViewById(R.id.toolbar);
        mRegistroDB = RegistroDB.getInstance(mContext);

        mCompactCalendarView = (CompactCalendarView) getActivity().findViewById(R.id.calendar);
        mCompactCalendarView.setVisibility(View.VISIBLE);
        mCompactCalendarView.setLocale(TimeZone.getTimeZone("Europe/Rome"), Locale.ITALIAN);
        mCompactCalendarView.setUseThreeLetterAbbreviation(true);
        mCompactCalendarView.setListener(this);
        mCompactCalendarView.shouldSelectFirstDayOfMonthOnScroll(false);

        addFAB.setClosedOnTouchOutside(true);

        verificaFAB.setOnClickListener(v -> startActivity(new Intent(mContext, AddEventActivity.class).putExtra("type", "Verifica").putExtra("time", mDate.getTime())));
        eserciziFAB.setOnClickListener(v -> startActivity(new Intent(mContext, AddEventActivity.class).putExtra("type", "Compiti").putExtra("time", mDate.getTime())));
        altroFAB.setOnClickListener(v -> startActivity(new Intent(mContext, AddEventActivity.class).putExtra("type", "Altro").putExtra("time", mDate.getTime())));

        adapter = new AgendaAdapter(mContext, place_holder, mRegistroDB);
        adapter.setItemClickListener(this);
        recycler.setLayoutManager(new LinearLayoutManager(mContext));
        recycler.setAdapter(adapter);

        prepareDate(true);
        mCompactCalendarView.setCurrentDate(mDate);

        updateDB();
    }

    private void prepareDate(boolean predictNextDay) {
        mDate = new Date();

        Calendar cal = toCalendar(mDate);

        if (predictNextDay) {
            boolean isOrarioScolastico = cal.get(Calendar.HOUR_OF_DAY) < 14;
            if (cal.get(Calendar.DAY_OF_WEEK) == Calendar.SATURDAY && !isOrarioScolastico) {
                cal.add(Calendar.DATE, 2);
            } else if (cal.get(Calendar.DAY_OF_WEEK) == Calendar.SUNDAY) {
                cal.add(Calendar.DATE, 1);
            } else if (!isOrarioScolastico)
                cal.add(Calendar.DATE, 1);
        }
        cal.set(Calendar.HOUR_OF_DAY, 0);
        cal.set(Calendar.MINUTE, 0);
        cal.set(Calendar.SECOND, 0);
        cal.set(Calendar.MILLISECOND, 0);

        mDate = cal.getTime();
    }

    private void fetchEvents() {
        events.clear();
        events.addAll(mRegistroDB.getAllEvents());
    }

    private void updateAdapter() {
        setAdapterEvents(mRegistroDB.getAllEvents(mDate.getTime()));
    }

    private void updateCalendar() {
        mCompactCalendarView.removeAllEvents();
        fetchEvents();
        mCompactCalendarView.addEvents(convertEvents(events));
        mCompactCalendarView.invalidate();

    }

    private Calendar toCalendar(Date date) {
        Calendar cal = Calendar.getInstance();
        cal.setTime(date);
        return cal;
    }

    private void updateDB() {
        new SpiaggiariApiClient(mContext)
                .getEvents(0L, Long.MAX_VALUE)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(events -> {
                    Log.d(TAG, "Scaricati " + events.size() + " eventi");
                    mRegistroDB.addEvents(events);

                    updateCalendar();
                    updateAdapter();

                }, error -> {
                    error.printStackTrace();
                    Snackbar.make(mCoordinatorLayout, error.getLocalizedMessage(), Snackbar.LENGTH_LONG).show();
                });
    }

    @Override
    public void onDayClick(Date dateClicked) {
        mDate = dateClicked;
        updateAdapter();
    }

    @Override
    public void onMonthScroll(Date firstDayOfNewMonth) {
        mDate = firstDayOfNewMonth;
        setTitleSubtitle(firstDayOfNewMonth);
    }

    private void setTitleSubtitle(Date d) {
        TransitionManager.beginDelayedTransition(mToolbar, new ChangeText().setChangeBehavior(ChangeText.CHANGE_BEHAVIOR_IN));
        mToolbar.setTitle(WordUtils.capitalizeFully(month.format(d)));
        mToolbar.setSubtitle(WordUtils.capitalizeFully(year.format(d)));
    }

    private void setAdapterEvents(List<AdvancedEvent> events) {
        adapter.clear();
        adapter.addAll(events);
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        inflater.inflate(R.menu.agenda, menu);
        super.onCreateOptionsMenu(menu, inflater);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == R.id.today) {
            prepareDate(false);
            mCompactCalendarView.setCurrentDate(mDate);
            setTitleSubtitle(mDate);

            setAdapterEvents(mRegistroDB.getAllEvents(mDate.getTime()));
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onDetach() {
        super.onDetach();
        mRegistroDB.close();
        mToolbar.setSubtitle("");
    }

    @Override
    public void onResume() {
        super.onResume();
        setTitleSubtitle(mDate);
        updateCalendar();
        updateAdapter();
    }

    @Override
    public void onAgendaItemClicked(AdvancedEvent e) {
        AgendaBS bottomSheetAgenda = new AgendaBS();
        bottomSheetAgenda.setEvent(e);
        bottomSheetAgenda.show(getChildFragmentManager(), "dialog");
    }

    @Override
    public void onBottomSheetItemClicked(int position, AdvancedEvent event) {
        String head = getSubjectNameOrProfessorName(event, mRegistroDB);
        switch (position) {
            case 0:
                if (mRegistroDB.isCompleted(event.getId()))
                    mRegistroDB.setUncompleted(event.getId());
                else mRegistroDB.setCompleted(event.getId());
                updateAdapter();
                break;
            case 1:
                Intent sendIntent = new Intent();
                sendIntent.setAction(Intent.ACTION_SEND);
                sendIntent.putExtra(Intent.EXTRA_TEXT, eventToString(event, head));
                sendIntent.setType("text/plain");
                startActivity(Intent.createChooser(sendIntent, getString(R.string.share_with)));
                break;
            case 2:
                addEventToCalendar(mContext, event);
                break;
            case 3:
                android.content.ClipboardManager clipboard = (android.content.ClipboardManager) mContext.getSystemService(Context.CLIPBOARD_SERVICE);
                android.content.ClipData clip = android.content.ClipData.newPlainText("Evento copiato", eventToString(event, head));
                clipboard.setPrimaryClip(clip);

                break;
            case 4:
                mRegistroDB.archive(event.getId());
                updateAdapter();
                updateCalendar();
                break;
        }
    }
}

File: app/src/main/java/com/sharpdroid/registroelettronico/fragments/FragmentAgenda.kt
package com.sharpdroid.registroelettronico.fragments

import android.arch.lifecycle.MediatorLiveData
import android.arch.lifecycle.Observer
import android.arch.lifecycle.ViewModelProviders
import android.content.Intent
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v7.widget.LinearLayoutManager
import android.util.Log
import android.view.*
import com.github.sundeepk.compactcalendarview.CompactCalendarView
import com.sharpdroid.registroelettronico.BuildConfig
import com.sharpdroid.registroelettronico.R
import com.sharpdroid.registroelettronico.activities.AddEventActivity
import com.sharpdroid.registroelettronico.adapters.AgendaAdapter
import com.sharpdroid.registroelettronico.api.cloud.v1.Cloud
import com.sharpdroid.registroelettronico.database.entities.LocalAgenda
import com.sharpdroid.registroelettronico.database.entities.RemoteAgendaInfo
import com.sharpdroid.registroelettronico.database.entities.SuperAgenda
import com.sharpdroid.registroelettronico.database.pojos.LocalAgendaPOJO
import com.sharpdroid.registroelettronico.database.room.DatabaseHelper
import com.sharpdroid.registroelettronico.fragments.bottomSheet.AgendaBS
import com.sharpdroid.registroelettronico.utils.Account
import com.sharpdroid.registroelettronico.utils.Metodi
import com.sharpdroid.registroelettronico.utils.Metodi.capitalizeEach
import com.sharpdroid.registroelettronico.utils.Metodi.convertEvents
import com.sharpdroid.registroelettronico.utils.Metodi.eventToString
import com.sharpdroid.registroelettronico.utils.Metodi.isEventTest
import com.sharpdroid.registroelettronico.utils.Metodi.pushDatabase
import com.sharpdroid.registroelettronico.utils.Metodi.updateAgenda
import com.sharpdroid.registroelettronico.utils.add
import com.sharpdroid.registroelettronico.utils.flat
import com.sharpdroid.registroelettronico.viewModels.AgendaViewModel
import com.transitionseverywhere.ChangeText
import com.transitionseverywhere.TransitionManager
import kotlinx.android.synthetic.main.activity_main.*
import kotlinx.android.synthetic.main.app_bar_main.*
import kotlinx.android.synthetic.main.fragment_agenda.*
import java.text.SimpleDateFormat
import java.util.*

// DONE: 19/01/2017 Aggiungere eventi all'agenda
// DONE: 19/01/2017 Aggiungere eventi dell'agenda nel calendario del telefono

class FragmentAgenda : Fragment(), CompactCalendarView.CompactCalendarViewListener, AgendaAdapter.AgendaClickListener, AgendaBS.Listener {
    private var month = SimpleDateFormat("MMMM", Locale.getDefault())
    private var year = SimpleDateFormat("yyyy", Locale.getDefault())
    internal var agenda = SimpleDateFormat("yyyyMMdd", Locale.getDefault())

    private val adapter by lazy { AgendaAdapter(place_holder) }
    private var mDate: Date = Date()
    private val events = ArrayList<Any>()
    private val local = ArrayList<LocalAgendaPOJO>()
    private val remote = ArrayList<SuperAgenda>()

    override fun onCreateView(inflater: LayoutInflater,
                              container: ViewGroup?, savedInstanceState: Bundle?): View? {
        setHasOptionsMenu(true)
        return inflater.inflate(R.layout.fragment_agenda, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        with(activity.calendar) {
            visibility = View.VISIBLE
            setLocale(TimeZone.getTimeZone("Europe/Rome"), Locale.ITALIAN)
            setUseThreeLetterAbbreviation(true)
            setListener(this@FragmentAgenda)
            shouldSelectFirstDayOfMonthOnScroll(false)


            //already been here
            if (savedInstanceState != null) {
                mDate = Date(savedInstanceState.getLong(INTENT_DATE))
            }
            //coming from a single item event notification
            else if (arguments != null &&
                    arguments.containsKey(INTENT_DATE)) {

                mDate = Date(arguments[INTENT_DATE] as Long)
                Log.d(this@FragmentAgenda.tag, mDate.toGMTString())
            } else {
                prepareDate(true)
            }
            setCurrentDate(mDate)
            updateAdapter()
            updateCalendar()
        }

        Log.d("FragmentAgenda", mDate.toString())

        with(activity) {
            fab_big_add.visibility = View.VISIBLE

            fab_big_add.setClosedOnTouchOutside(true)
            fab_mini_verifica.setOnClickListener {
                startActivity(Intent(context, AddEventActivity::class.java)
                        .putExtra("type", "Verifica").putExtra("time", mDate.time))
                fab_big_add.close(true)
            }
            fab_mini_esercizi.setOnClickListener {
                startActivity(Intent(context, AddEventActivity::class.java)
                        .putExtra("type", "Compiti")
                        .putExtra("time", mDate.time))
                fab_big_add.close(true)
            }
            fab_mini_altro.setOnClickListener {
                startActivity(Intent(context, AddEventActivity::class.java)
                        .putExtra("type", "Altro")
                        .putExtra("time", mDate.time))
                fab_big_add.close(true)
            }
        }
        adapter.setItemClickListener(this)
        recycler.layoutManager = LinearLayoutManager(context)
        recycler.adapter = adapter

        if (savedInstanceState == null)
            download()

        val viewModel = ViewModelProviders.of(this)[AgendaViewModel::class.java]

        val mediator = MediatorLiveData<List<Any>>()


        //TODO: Use mediator but only observe when changes happen
        mediator.addSource(viewModel.getLocal(Account.with(context).user), {
            local.clear()
            local.addAll(it.orEmpty())

            events.clear()
            events.addAll(local)
            events.addAll(remote)
            mediator.value = null
        })

        mediator.addSource(viewModel.getRemote(Account.with(context).user), {
            remote.clear()
            remote.addAll(it?.map {
                SuperAgenda(it.event, it.isCompleted(), it.isTest())
            }.orEmpty())

            events.clear()
            events.addAll(local)
            events.addAll(remote)
            mediator.value = null
        })

        mediator.observe(this, Observer {
            updateAdapter()
            updateCalendar()
        })


        val profile = Account.with(context).user

        Cloud.api.pullLocal(profile).subscribe({
            DatabaseHelper.database.eventsDao().insertLocal(it)
        }, { Log.e("Cloud", it.localizedMessage) })

        Cloud.api.pullRemoteInfo(profile).subscribe({
            DatabaseHelper.database.eventsDao().insertRemoteInfo(it)
        }, { Log.e("Cloud", it.localizedMessage) })
    }

    private fun download() {
        updateAgenda(context)
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putLong(INTENT_DATE, mDate.time)
    }

    override fun onResume() {
        super.onResume()
        activity.calendar.visibility = View.VISIBLE
        setTitleSubtitle(mDate)
    }

    override fun onStop() {
        super.onStop()

        pushDatabase(context)
    }

    private fun prepareDate(predictNextDay: Boolean) {
        mDate = Date()

        val cal = toCalendar(mDate)

        if (predictNextDay) {
            val isOrarioScolastico = cal.get(Calendar.HOUR_OF_DAY) < 14
            if (cal.get(Calendar.DAY_OF_WEEK) == Calendar.SATURDAY && !isOrarioScolastico) {
                cal.add(Calendar.DATE, 2)
            } else if (cal.get(Calendar.DAY_OF_WEEK) == Calendar.SUNDAY) {
                cal.add(Calendar.DATE, 1)
            } else if (!isOrarioScolastico)
                cal.add(Calendar.DATE, 1)
        }

        mDate = cal.time.flat()
    }

    private fun fetch(currentDate: Boolean?): List<Any> {
        return if (currentDate == true)
            events.filter {
                when (it) {
                    is SuperAgenda -> return@filter it.agenda.end.time in mDate.time until mDate.add(Calendar.HOUR_OF_DAY, 24).time && it.agenda.start.time in mDate.time until mDate.add(Calendar.HOUR_OF_DAY, 24).time
                    is LocalAgendaPOJO -> return@filter it.event.day in mDate.time until mDate.add(Calendar.HOUR_OF_DAY, 24).time
                }
                true
            }
        else events
    }

    private fun updateAdapter() {
        setAdapterEvents(fetch(true))
    }

    private fun updateCalendar() {
        with(activity.calendar) {
            removeAllEvents()
            addEvents(convertEvents(events))
            invalidate()
        }
    }

    private fun toCalendar(date: Date): Calendar {
        val cal = Calendar.getInstance()
        cal.time = date
        return cal
    }

    override fun onDayClick(dateClicked: Date) {
        mDate = dateClicked
        updateAdapter()
    }

    override fun onMonthScroll(firstDayOfNewMonth: Date) {
        mDate = firstDayOfNewMonth
        setTitleSubtitle(firstDayOfNewMonth)
    }

    private fun setTitleSubtitle(d: Date?) {
        TransitionManager.beginDelayedTransition(activity.toolbar, ChangeText().setChangeBehavior(ChangeText.CHANGE_BEHAVIOR_IN))
        activity.toolbar.title = capitalizeEach(month.format(d))
        activity.toolbar.subtitle = capitalizeEach(year.format(d))
    }

    private fun setAdapterEvents(events: List<Any>) {
        adapter.clear()
        adapter.addAll(events)
    }

    override fun onCreateOptionsMenu(menu: Menu?, inflater: MenuInflater?) {
        inflater!!.inflate(R.menu.agenda, menu)
        super.onCreateOptionsMenu(menu, inflater)
    }

    override fun onOptionsItemSelected(item: MenuItem?): Boolean {
        if (item!!.itemId == R.id.today) {
            prepareDate(false)
            activity.calendar.setCurrentDate(mDate)
            setTitleSubtitle(mDate)

            setAdapterEvents(fetch(true))
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onDetach() {
        super.onDetach()
        activity.toolbar.subtitle = ""
    }


    override fun onAgendaItemClicked(e: Any) {
        val bottomSheetAgenda = AgendaBS()
        bottomSheetAgenda.setEvent(e)
        bottomSheetAgenda.show(childFragmentManager, "dialog")
    }

    override fun onBottomSheetItemClicked(position: Int, e: Any) {
        if (e is SuperAgenda) {
            when (position) {
                0 -> {
                    val found: RemoteAgendaInfo? = e.agenda.getInfo()
                    if (found != null) {
                        found.completed = !found.completed
                        DatabaseHelper.database.eventsDao().update(found)
                    } else {
                        DatabaseHelper.database.eventsDao().insertRemoteInfo(RemoteAgendaInfo(e.agenda.id, true, false, isEventTest(e)))
                    }

                    e.completed = !e.completed
                }
                1 -> {
                    val found = e.agenda.getInfo()
                    if (found != null) {
                        found.test = !found.test
                        DatabaseHelper.database.eventsDao().update(found)
                    } else {
                        DatabaseHelper.database.eventsDao().insertRemoteInfo(RemoteAgendaInfo(e.agenda.id, false, false, !isEventTest(e)))
                    }
                }
                2 -> {
                    val sendIntent = Intent()
                    sendIntent.action = Intent.ACTION_SEND
                    sendIntent.putExtra(Intent.EXTRA_TEXT, eventToString(e))
                    sendIntent.type = "text/plain"
                    startActivity(Intent.createChooser(sendIntent, getString(R.string.share_with)))
                }
                3 -> {
                    val found: RemoteAgendaInfo? = e.agenda.getInfo()
                    if (found != null) {
                        found.archived = true
                        DatabaseHelper.database.eventsDao().update(found)
                    } else {
                        DatabaseHelper.database.eventsDao().insertRemoteInfo(RemoteAgendaInfo(e.agenda.id, false, true, isEventTest(e)))
                    }
                }
            }
        } else if (e is LocalAgenda) { //no need to invalidate since cache is used only for RemoteAgenda
            when (position) {
                0 -> {
                    e.completed_date = if (e.completed_date != 0L) 0L else System.currentTimeMillis()
                    DatabaseHelper.database.eventsDao().insertLocal(e)
                }
                1 -> {
                    e.type = if (e.type == "verifica") "altro" else "verifica"
                    DatabaseHelper.database.eventsDao().insertLocal(e)
                }
                2 -> {
                    val sendIntent = Intent()
                    sendIntent.action = Intent.ACTION_SEND
                    sendIntent.putExtra(Intent.EXTRA_TEXT, Metodi.complex.format(e.day) + "\n" + e.title)
                    sendIntent.type = "text/plain"
                    startActivity(Intent.createChooser(sendIntent, getString(R.string.share_with)))
                }
                3 -> {
                    e.archived = !e.archived
                    DatabaseHelper.database.eventsDao().insertLocal(e)
                }
            }
        }
    }

    companion object {
        val INTENT_DATE = "date"
    }
}
